### DISCUSSION TITLE
Async issues

### DISCUSSION THREAD

**Discussion by fundez** [MarimerLLC/csla#4382] [Questions] [answered]
This is probably more of a blazor issue than CSLA, but since the code revolves around csla I figured I'd ask here.

I'm creating an business object with this call.  fagj = await applicationContext.GetRequiredService<IDataPortal<FezWeb.BizObj.FA_FAGJ>>().CreateAsync();

The create calls a common object to retrieve a journal # and the value is assigned to the journalNo property. The issue we're running into is the code returns before the command object completes it's process. The next line of code relies on the journal property being filled in. Adding a Task.Delay between the lines of code solves the issue. Is their a better approach to waiting for processes to complete.

Here is my create code 

  using (BypassPropertyChecks)
  {
  
      LoadProperty(CO_DetailsListProperty, await DalCO_DetailsList.CreateChildAsync());
      LoadProperty(FR_GiftsListProperty, await DalFR_GiftsList.CreateChildAsync());
      LoadProperty(CO_IntermediatesListProperty, await DalCO_IntermediatesList.CreateChildAsync());
      LoadProperty(CO_HeaderGenericsListProperty, await DalCO_HeaderGenericsList.CreateChildAsync());
      LoadProperty(CO_MemorizedListProperty, await DalCO_MemorizedList.CreateChildAsync());
      LoadProperty(PO_POHeadersListProperty, await DalPO_POHeadersList.CreateChildAsync());
      LoadProperty(CB_CBHeadersListProperty, await DalCB_CBHeadersList.CreateChildAsync());
      LoadProperty(FR_PledgeSchedulesListProperty, await DalFR_PledgeSchedulesList.CreateChildAsync());
      var args = new DataPortalHookArgs();
    SetDefaults();
}
Thanks,

Mike

**Comment by fundez** [comment]
adding the code below within the create solves it, but not sure if this is the best approach.

Task T = Task.Run(async () => await SetDefaults());
T.Wait();

**Comment by StefanOssendorf** [comment]
Please post your whole create method with declaration.
And the Csla version you are using.

**Comment by Bowman74** [comment]
I'd guess without seeing the rest of the project this issue is almost certainly because the async code is not properly written to flow up and down the stack. I'd guess somewhere in the stack there is something like an async void method or an async Task method called without an await operator. As soon as such a stack encounters "real" async code it is going to immediately return.

The solution given was to turn the async code into synchronous code which further gives me the feeling that this is the case.

What is the solution if I'm right? That would completely be dependent on the overall codebase and what is trying to be done. To me this looks like a common issue of not having the asynchronous code completely flow up and down the call stack and nothing to do with CSLA per se.

**Reply by StefanOssendorf** [reply]
I too guess there's an async void at the top. But can't say for sure without the portal method declaration

**Comment by fundez** [comment, marked as answer]
We're using net 9.0. the create is a void, let me try changing that, can't believe i missed that. "private async void DataPortal_CreateAsync([Inject] Csla.IDataPortal<CO_Headers> Dal, "

Let me try changing it.

**Comment by fundez** [comment]
That solved it, can't believe I missed something so simple.

Thanks for your help.

**Reply by StefanOssendorf** [reply]
Glad we could help. 

On my own behalf: I can recommend this generator ;)
https://github.com/StefanOssendorf/Csla.DataPortalExtensions 😇
